python-package/lightgbm/basic.py

```diff
@@ -712,6 +712,8 @@ def _lazy_init(self, data, label=None, reference=None,
             self.__init_from_csc(data, params_str, ref_dataset)
         elif isinstance(data, np.ndarray):
             self.__init_from_np2d(data, params_str, ref_dataset)
+        elif isinstance(data, list) and len(data) > 0 and all(isinstance(x, np.ndarray) for x in data):
```

Comment:
Speaking about Python side, and except the remark about the memory saving, maybe not only `list` of numpy arrays, but any indexable and iterable object?
```
isinstance(data, list)
```

Reply:
We might use [collections.abc.Sequence](https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence), but it is supported only in python >3.3.

Maybe `instance(data, (list, tuple))` is going to be good enough? Or just leave it as is.

Reply:
To be honest, I'm worrying most of all about numpy array. So, `list, tuple,  np.ndarray` is enough, I suppose.

Reply:
Is it possible to create `ndarray` of `ndarray`-s?

Reply:
Seems to be that not. You're right. I thought that it's possible.
Then `list` and `tuple` will be enough.